### PR TITLE
enhance(StoragePool): add owner references to StoragePool related CRs for garbage collection

### PR DESCRIPTION
**Pull request by ashishranjan738**
Signed-off-by: Ashish Ranjan <ashishranjan738@gmail.com>

Currently, the StoragePool related CRS don't have owner references which can lead to stale resources in K8s when the hierarchy of the storage pools breaks. The hierarchy followed by the StoragePools is ` SPC -> CSP -> Pool Deployment -> SP `. This issue can be solved by using the power of k8s garbage collection. For information regarding this refer to this https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/ .

**What this PR does / why we need it**:
-> This PR adds owner references to StoragePool and PoolDeployment CR's which can be used by the garbage collector for collecting stale objects.
-> Adding owner references to Resources will delete all the orphaned objects when the parent is deleted which will help to fight the stale situation.
-> The owner of SP is Pool Deployment and owner of Pool Deployment is CSP.
-> Example:
```apiVersion: v1
items:
- apiVersion: openebs.io/v1alpha1
  kind: StoragePool
  metadata:
    creationTimestamp: "2019-01-07T06:05:16Z"
    generation: 1
    labels:
      kubernetes.io/hostname: 127.0.0.1
      openebs.io/cas-template-name: cstor-pool-create-default-0.8.0
      openebs.io/cas-type: cstor
      openebs.io/cstor-pool: cstor-sparse-pool-4aer
      openebs.io/storage-pool-claim: cstor-sparse-pool
      openebs.io/version: 0.8.0
    name: cstor-sparse-pool-4aer
    ownerReferences:
    - apiVersion: openebs.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Deployment
      name: cstor-sparse-pool-4aer
      uid: 341365f1-1242-11e9-aec9-b4b686bd0cff
    resourceVersion: "584"
    selfLink: /apis/openebs.io/v1alpha1/storagepools/cstor-sparse-pool-4aer
    uid: 3416a13d-1242-11e9-aec9-b4b686bd0cff
    .
    .

---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
  creationTimestamp: "2019-01-07T06:05:16Z"
  generation: 1
  labels:
    app: cstor-pool
    openebs.io/cas-template-name: cstor-pool-create-default-0.8.0
    openebs.io/cstor-pool: cstor-sparse-pool-4aer
    openebs.io/storage-pool-claim: cstor-sparse-pool
    openebs.io/version: 0.8.0
  name: cstor-sparse-pool-4aer
  namespace: openebs
  ownerReferences:
  - apiVersion: openebs.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: CStorPool
    name: cstor-sparse-pool-4aer
    uid: 3410d067-1242-11e9-aec9-b4b686bd0cff
  resourceVersion: "614"
```
    
**NOTE**:
`blockOwnerDeletion`: Making it true will delete the owner object only after the child is deleted.
`controller`: Making it true will make the owner reference points to the managing controller, in other words, the owner. And also, it is the instruction for Garbage Collector's behavior related to the object and its owner. If controller: false, Garbage Collector manages the object as an object without an owner, for example, allows to delete it freely.

**Steps to verify**:
-> Provision CStor Pool.
-> Check the metadata of StoragePool and PoolDeployment of StoragePool.

### CODE REVIEW

File: pkg/install/v1alpha1/cstor_pool.go [hunk 1]
```diff
@@ -115,9 +115,9 @@ spec:
   post: |
     {{- jsonpath .JsonResult "{.metadata.name}" | trim | addTo "putcstorpoolcr.objectName" .TaskResult | noop -}}
     {{- jsonpath .JsonResult "{.metadata.uid}" | trim | addTo "putcstorpoolcr.objectUID" .TaskResult | noop -}}
-    {{- jsonpath .JsonResult "{.metadata.labels.kubernetes\\.io/hostname}" | trim | addTo "putcstorpoolcr.nodeName" .TaskResult | noop -}}
-  task: |-
-    {{- $diskDeviceIdList:= .Storagepool.diskDeviceIdList }}
+    {{- $nodeUidList := jsonpath .JsonResult "pkey=nodeUid,{.metadata.labels.kubernetes\\.io/hostname}={.metadata.uid} {.metadata.name};" | trim | default "" | splitList ";" -}}
```
Comment (AmitKumarDas) [inline pkg/install/v1alpha1/cstor_pool.go]: What are we trying to do here? I guess this logic is over engineered? Can you please come up with a simpler alternative?

Comment (ashishranjan738) [reply pkg/install/v1alpha1/cstor_pool.go]: done

File: pkg/install/v1alpha1/cstor_pool.go [hunk 2]
```diff
@@ -170,6 +171,13 @@ spec:
         app: cstor-pool
         openebs.io/version: {{ .CAST.version }}
         openebs.io/cas-template-name: {{ .CAST.castName }}
+      ownerReferences:
+      - apiVersion: openebs.io/v1alpha1
+        blockOwnerDeletion: true
+        controller: true
+        kind: CStorPool
```
Comment (AmitKumarDas) [inline pkg/install/v1alpha1/cstor_pool.go]: How can CSP be an owner when the PR talks of below relationship?
`SPC -> SP -> Pool Deployment -> CSP`

**Comment by AmitKumarDas** [comment]
@ashishranjan738 can you please provide better commit message to understand the changes?
The commit message as well as the PR is too terse to understand.

File: pkg/install/v1alpha1/cstor_pool.go (hunk 2 above)
Comment (ashishranjan738) [reply pkg/install/v1alpha1/cstor_pool.go]: Sry the relationship is SPC -> CSP -> Pool Deployment -> SP.

**Comment by sonasingh46** [comment]
@ashishranjan738 -- Can we add SPC as an owner for Pool Deployment and CSP.

**Comment by ashishranjan738** [comment]
> @ashishranjan738 -- Can we add SPC as an owner for Pool Deployment and CSP.

added

**Pull request merged**